### PR TITLE
coretasks: split AUTHENTICATE token in 400-byte chunks

### LINKED ISSUE #975: AUTHENTICATE arguments are not split.
When using a very long password, the authstring given with AUTHENTICATE exceeds 400 characters, and thus should be splitted, according to [the specification](http://ircv3.net/specs/extensions/sasl-3.1.html#the-authenticate-command)

Config file:

```
[core]
nick = Sopel
host = 0.0.0.0
use_ssl = false
port = 34827
owner = me
channels = 
auth_username = foo
auth_password = barbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbar
auth_method = sasl
```

Sopel session log:

```
C: CAP LS 302
S: CAP * LS :sasl
C: NICK Sopel
C: USER sopel +iw Sopel :Sopel: http://sopel.chat
C: CAP REQ multi-prefix
S: CAP Sopel NAK :multi-prefix
C: CAP REQ sasl
S: CAP Sopel ACK :sasl
C: AUTHENTICATE PLAIN
S: AUTHENTICATE +
C: AUTHENTICATE Zm9vAGZvbwBiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiY
```

Expected log:

```
C: CAP LS 302
S: CAP * LS :sasl
C: NICK Sopel
C: USER sopel +iw Sopel :Sopel: http://sopel.chat
C: CAP REQ multi-prefix
S: CAP Sopel NAK :multi-prefix
C: CAP REQ sasl
S: CAP Sopel ACK :sasl
C: AUTHENTICATE PLAIN
S: AUTHENTICATE +
C: AUTHENTICATE :Zm9vAGZvbwBiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJi
C: AUTHENTICATE :YXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJiYXJi
C: AUTHENTICATE :YXJiYXJiYXI=
```

Version: 047a99ed2d3f76c7d54e5bacaac1c1d38d1b743c (Git master at the time I am writing this)

### PR DESCRIPTION
**Pull request by Exirel**
Fix #975

Link to the [AUTHENTICATE specification](https://ircv3.net/specs/extensions/sasl-3.1.html#the-authenticate-command).

> The response is encoded in Base64 (RFC 4648), then split to 400-byte chunks, and each chunk is sent as a separate AUTHENTICATE command. Empty (zero-length) responses are sent as AUTHENTICATE +. If the last chunk was exactly 400 bytes long, it must also be followed by AUTHENTICATE + to signal end of response.

### CODE REVIEW

**Comment by HumorBaby** [comment]
So... I couldn't get a services spun (Atheme 7.2) up to accept a `SASL PLAIN` password with > 300 characters (base64 encodes to 400-bytes). Too many moving parts for a weary brain... :weary: 

So checked the `raw` log instead (and assumed that any *sane* services would authenticate if correct creds) *did show* that the AUTHENTICATE response was sent as required in the specification.

I will admit, I'm turned off of testing/debugging IRC protocol for a while :trollface:

File: sopel/coretasks.py [hunk 1]
```diff
@@ -539,6 +539,40 @@ def recieve_cap_ack_sasl(bot):
...
+    last_size = 0
+
+    for i in range(0, len(payload), chunk_size):
+        offset = i + chunk_size
+        chunk = payload[i:offset]
+        last_size = len(chunk)
+        bot.write(('AUTHENTICATE', chunk))
+
+    # send empty (+) AUTHENTICATE command when last chunk's size is 400
+    if last_size == chunk_size:
```
Comment (kwaaak) [inline sopel/coretasks.py]: What about `if len(payload) % chunk_size == 0`

File: sopel/coretasks.py [hunk 2]
```diff
@@ -539,6 +539,37 @@ def recieve_cap_ack_sasl(bot):
...
 
 
+def send_authenticate(bot, token):
+    """Send ``AUTHENTICATE`` command to server with the given ``token``.
+
+    :param bot: instance of IRC bot that must authenticate
+    :param str token: authentication token
+
+    In case the ``token`` is more than 400 bytes, we need to split it and send
+    as many ``AUTHENTICATE`` command as needed. If the last chunk is 400 bytes
```
Comment (dgw) [inline sopel/coretasks.py]:
```suggestion
    as many ``AUTHENTICATE`` commands as needed. If the last chunk is 400 bytes
```

I'm nitpicking docstrings, as usual. :joy_cat:

**Comment by Exirel** [comment]
@dgw and @kwaaak I've modified as requested, so give it a last review and let's merge this!

**Comment by dgw** [comment]
This needs to be rebased on the 6.6.x branch. Right now it would pull in changes from `master` that we don't want to release in 6.6.2. :)

**Comment by dgw** [comment]
(Also the commit shortmessage is slightly too long… When rebasing, you should be able to make it fit on one line on GitHub by editing it to: "coretasks: split AUTHENTICATE token in 400-byte chunks (fix #975)"

**Comment by Exirel** [comment]
Rebased from 6.6.x as expected. All is good now.

**Pull request merged**